### PR TITLE
ChibiOS: added microSD param storage options

### PR DESCRIPTION
**Pull request by tridge**
This is useful for boards with microSD and without a flash layout that allows for flash storage

### CODE REVIEW

File: libraries/AP_Arming/AP_Arming.cpp [hunk 1]
```diff
@@ -541,6 +541,19 @@ bool AP_Arming::board_voltage_checks(bool report)
     return true;
 }
 
+/*
+  check base system operations
+ */
+bool AP_Arming::system_checks(bool report)
+{
+    if ((checks_to_perform & ARMING_CHECK_ALL) || (checks_to_perform & ARMING_CHECK_SYSTEM)) {
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp]: This should be `check_enabled(ARMING_CHECK_SYSTEM)`

File: libraries/AP_HAL_ChibiOS/Storage.cpp [hunk 2]
```diff
@@ -46,8 +60,37 @@ void Storage::_storage_open(void)
...
+#ifdef STORAGE_FLASH_PAGE
     // load from storage backend
     _flash_load();
+#endif
+    
+#ifdef USE_POSIX
+    // allow for fallback to microSD based storage
+    sdcard_init();
+    
+    log_fd = open(HAL_STORAGE_FILE, O_RDWR|O_CREAT);
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/Storage.cpp]: Using a temporary variable here and assigning it down the bottom would remove a few lines and close off future potential bus where log_fd wasn't reset to -1 on error.

File: libraries/AP_HAL_ChibiOS/Storage.cpp [hunk 3]
```diff
@@ -119,8 +165,27 @@ void Storage::_timer_tick(void)
     } 
 #endif
 
+#ifdef USE_POSIX
+    if (using_filesystem && log_fd != -1) {
+        uint32_t offset = CH_STORAGE_LINE_SIZE*i;
+        if (lseek(log_fd, offset, SEEK_SET) != offset) {
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/Storage.cpp]: Should we be checking for `-1`s being returned here?

File: libraries/AP_Arming/AP_Arming.cpp [hunk 4]
```diff
@@ -573,6 +587,14 @@ bool AP_Arming::arm_checks(uint8_t method)
         }
     }
 
+    // check system health on arm as well as prearm
+    if ((checks_to_perform & ARMING_CHECK_ALL) ||
```
Comment (WickedShell) [inline libraries/AP_Arming/AP_Arming.cpp]: This is a redundant guard, system_checks already does the check against `checks_to_preform`

Comment (tridge) [reply libraries/AP_Arming/AP_Arming.cpp]: that's what I thought, but it doesn't work. The pre-arm checks are cached in copter. So if the microSD fails before arming (or is removed), it allows arming with bad storage.

File: libraries/AP_HAL_ChibiOS/Storage.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Storage.cpp]: it ends up no shorter, as log_fd also is needed as a check for having already tried to open

File: libraries/AP_HAL_ChibiOS/Storage.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Storage.cpp]: we're checking == offset, which is stronger than checking for not getting -1

File: libraries/AP_Arming/AP_Arming.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_Arming/AP_Arming.cpp]: ideally we'd have a mask of checks that should be cached, and which ones should be specifically checked again

Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp]: I thought I removed the pre-arm-check-caching. 
7b3e111f85d936db047d81ba5a49d1a22fb0667a

File: libraries/AP_HAL_ChibiOS/Storage.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_HAL_ChibiOS/Storage.cpp]: But we are not setting log_fd to -1 etc etc to indicate everything is stuffed - sorry, I wasn't clear.

Comment (tridge) [reply libraries/AP_HAL_ChibiOS/Storage.cpp]: ahh, I want to allow for temporary failures. I don't want a single failure to cause a problem.

File: libraries/AP_Arming/AP_Arming.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_Arming/AP_Arming.cpp]: it didn't seem to work when I tested today. If I pulled the microSD out a few seconds before arming it allowed me to arm, and ChibiOS::Storage::healthy() wasn't called

**Pull request merged**